File: python/day04.py
import re

# why this does not work
# pattern = re.compile('\[\d{4}-(?P<day>\d{2}-\d{2}) (?P<hour>\d{2}):(?P<minute>\d{2})\] (?:Guard #(\d+) begins shift)|(falls asleep)|(wakes up)')
# while this does?
pattern = re.compile('\[\d{4}-(?P<day>\d{2}-\d{2}) (?P<hour>\d{2}):(?P<minute>\d{2})\] (?:Guard #(?P<id>\d+) begins shift)?(?P<asleep>falls asleep)?(?P<awake>wakes up)?')

# how much easier would be to "compile" a func that produces a string given some inputs? it would be a different approach that re and maybe doable in Nim! (because of programmability)

test_input = """[1518-11-01 00:00] Guard #10 begins shift
[1518-11-01 00:05] falls asleep
[1518-11-01 00:25] wakes up
[1518-11-01 00:30] falls asleep
[1518-11-01 00:55] wakes up
[1518-11-01 23:58] Guard #99 begins shift
[1518-11-02 00:40] falls asleep
[1518-11-02 00:50] wakes up
[1518-11-03 00:05] Guard #10 begins shift
[1518-11-03 00:24] falls asleep
[1518-11-03 00:29] wakes up
[1518-11-04 00:02] Guard #99 begins shift
[1518-11-04 00:36] falls asleep
[1518-11-04 00:46] wakes up
[1518-11-05 00:03] Guard #99 begins shift
[1518-11-05 00:45] falls asleep
[1518-11-05 00:55] wakes up"""
#123456789012345678901234567890
#         111111111122222222222
# I think I will need a regular expression here!

# processing test input
for line in test_input.split('\n'):
    print(line)
    match = pattern.search(line)
    if match is None:
        print('None')
    else:
        print({key: match.group(key) for key in 'day hour minute id asleep awake'.split()})


# I will have a set containing all shifts processed from input
class Shift:

    def __init__(self, id, date):
        self.id = id
        self.date = date
        # 0 index: asleep, 1 index: awake, 2: asleep, 3: awake, ...
        self.sleeps = list()
        self.events = 0
    
    def __repr__(self):
        return f"Shift(id={self.id}, date={self.date}, events={self.events}, sleeps={self.sleeps})"

    def add_sleep_awake(self, minute, asleep=True):
        self.sleeps.append(minute)
        if asleep:
            self.events += 1


# I will also have a dictionary with all guard history (with id as key)
class GuardHistory:

    def __init__(self, id):
        self.id = id
        self.minutes = [0 for i in range(60)]
    
    def __repr__(self):
        return f"Guard(id={self.id}, minutes={self.minutes})"

    def add_shift(self, shift):
        for event in range(shift.events):
            for m in range(shift.sleeps[2*event], shift.sleeps[2*event+1]):
                self.minutes[m] += 1
        return self



def process_input(text):
    shifts = set()
    guards = {}
    shift = None
    for line in text.split('\n'):
        if line == '':
            continue
        match = pattern.search(line)
        if match is None:
            raise ValueError('cannot process line:', line)
        if match.group('id'):
            # new shift, if there was a previous shift (if this is not the first one), update guard history
            if shift:
                shifts.add(shift)
                if shift.id in guards:
                    guards[shift.id].add_shift(shift)
                else:
                    guards[shift.id] = GuardHistory(shift.id).add_shift(shift)
            shift = Shift(id=match.group('id'), date=match.group('day'))
        elif match.group('asleep'):
            shift.add_sleep_awake(int(match.group('minute')))
        else:  # awake
            shift.add_sleep_awake(int(match.group('minute')), asleep=False)
        # update date if first date fell on day before
        if shift.date != match.group('day'):
            shift.date = match.group('day')
    return shifts, guards


shifts, guards = process_input(test_input)
print(shifts)
print(guards)

# Strategy 1 - test case
guards_by_minutes = {id: sum(guard.minutes) for id, guard in guards.items()}
print(guards_by_minutes)
guard_id = max(guards_by_minutes, key=guards_by_minutes.get)
print("Guard who spend the most minutes asleep", guard_id)
minutes = guards[guard_id].minutes
minute_max_asleep = max(range(len(minutes)), key=minutes.__getitem__)
print("Minute most asleep for guard", guard_id, ":", minute_max_asleep)
print("Solution: ", int(guard_id)*minute_max_asleep)

# Strategy 1 - real case
print("Real case:")
with open('./inputs/04.txt') as f:
    input = '\n'.join(sorted(f.readlines()))

shifts, guards = process_input(input)
guards_by_minutes = {id: sum(guard.minutes) for id, guard in guards.items()}
guard_id = max(guards_by_minutes, key=guards_by_minutes.get)
print("Guard who spend the most minutes asleep", guard_id)
minutes = guards[guard_id].minutes
minute_max_asleep = max(range(len(minutes)), key=minutes.__getitem__)
print("Minute most asleep for guard", guard_id, ":", minute_max_asleep)
print("Solution: ", int(guard_id)*minute_max_asleep)

# part 2
best_minute_by_guard = {id: max(range(len(guard.minutes)), key=guard.minutes.__getitem__) for id, guard in guards.items()}
best_guard = max(best_minute_by_guard, key=best_minute_by_guard.get)
best_minute = best_minute_by_guard[best_guard]
print("best guard: ", best_guard, "; its best minute: ", best_minute, "; their product: ", int(best_guard)*best_minute)

File: nim/day04.nim
# regular expressions in nim. there are (currently) 2 packages in stdlib: re and nre
# see: https://github.com/nim-lang/Nim/issues/7278
# let's try first re
import re, strutils

#                        0: day        1: hour 2: minute           3: id               4: asleep      5: awake   
let pattern = re"\[\d{4}-(\d{2}-\d{2}) (\d{2}):(\d{2})\] (?:Guard #(\d+) begins shift)?(falls asleep)?(wakes up)?"

var matches: array[6, string]

let test_input = """[1518-11-01 00:00] Guard #10 begins shift
[1518-11-01 00:05] falls asleep
[1518-11-01 00:25] wakes up
[1518-11-01 00:30] falls asleep
[1518-11-01 00:55] wakes up
[1518-11-01 23:58] Guard #99 begins shift
[1518-11-02 00:40] falls asleep
[1518-11-02 00:50] wakes up
[1518-11-03 00:05] Guard #10 begins shift
[1518-11-03 00:24] falls asleep
[1518-11-03 00:29] wakes up
[1518-11-04 00:02] Guard #99 begins shift
[1518-11-04 00:36] falls asleep
[1518-11-04 00:46] wakes up
[1518-11-05 00:03] Guard #99 begins shift
[1518-11-05 00:45] falls asleep
[1518-11-05 00:55] wakes up"""

# processing test input
for line in test_input.splitLines():
    echo line
    if match(line, pattern, matches):
      echo matches
# bizarre behaviour, last element of array is not reset when no match is given! ("wakes up" always appear after processing line 3!)

# I will do something different from python: a Table (similar to a python dict) of guards minutes
import tables, math

# not seeing in the documentation the fact that [A,B] refer to types of key, values of table
var guards = initTable[int, array[60, int]](1024)  # remember to set an adequate initial size!

var id, minute: int
var events: seq[int]
var minutes: array[60, int]

proc zero(minutes: var array[60, int]) =
  for i in minutes.low .. minutes.high:  # better way to zero again an array?
    minutes[i] = 0


proc update(minutes: var array[60, int], events: seq[int]) =
  for i in 0 ..< (events.len div 2):
    for j in events[2*i] ..< events[2*i + 1]:
      inc minutes[j]

minutes.update(@[5, 10])
echo minutes
minutes.update(@[5, 15])
echo minutes
zero(minutes)
echo minutes

proc process(text: string): Table[int, array[60, int]] =
  result = initTable[int, array[60, int]](1024)
  for line in text.splitLines():
    if not match(line, pattern, matches):
      echo "cannot understand this line: ", line
      continue
    if matches[3].len > 0:  # id: update guard table with last shift and create new shift
      if id != 0:  # first pass id is == 0
        # update guards
        if not guards.hasKey(id):
          zero(minutes)
          result[id] = minutes
        result[id].update(events)
      # create new shift data
      id = parseInt(matches[3])
      events = @[]
    else:
      events.add(parseInt(matches[2]))  # minute of other event

guards = process(test_input)

proc solve(guards: Table[int, array[60, int]]) =
  # guard with most minutes
  var idMax, totMins: int
  for id, minutes in guards:
    echo id, ": ", minutes.sum()
    if minutes.sum() > totMins:
      totMins = minutes.sum()
      idMax = id

  echo "guards with most minutes: ", idMax

  # minutes more often asleep
  var bestMinute, timesAsleep: int
  for minute in guards[idMax].low .. guards[idMax].high:
    if guards[idMax][minute] > timesAsleep:
      timesAsleep = guards[idMax][minute]
      bestMinute = minute

  echo "best minute: ", bestMinute
  echo "answer part 1: ", bestMinute*idMax

solve(guards)

# with real input:
import algorithm

const input = sorted(readFile("./inputs/04.txt").strip().splitLines(), system.cmp).join(sep="\n")

guards = process(input)
solve(guards)